In [1]:
import pandas as pd
df = pd.read_csv("bhp.csv")
df.head(10)

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   
5                Whitefield      2 BHK      1170.0   2.0   38.00    2   
6          Old Airport Road      4 BHK      2732.0   4.0  204.00    4   
7              Rajaji Nagar      4 BHK      3300.0   4.0  600.00    4   
8              Marathahalli      3 BHK      1310.0   3.0   63.25    3   
9                     other  6 Bedroom      1020.0   6.0  370.00    6   

   price_per_sqft  
0            3699  
1            4615  
2            4305  
3            6245  
4            4250  
5            3247  
6            7467  
7           18181  
8            4828  
9           36274

In [9]:
min_threshold, max_threshold = df['price_per_sqft'].quantile([0.001 , 0.999])
min_threshold , max_threshold

(1366.184, 50959.36200000098)

In [10]:
df[df.price_per_sqft < min_threshold]

location       size  total_sqft  bath  price  bhk  price_per_sqft
665    Yelahanka      3 BHK     35000.0   3.0  130.0    3             371
798        other  4 Bedroom     10961.0   4.0   80.0    4             729
1867       other  3 Bedroom     52272.0   2.0  140.0    3             267
2392       other  4 Bedroom      2000.0   3.0   25.0    4            1250
3934       other      1 BHK      1500.0   1.0   19.5    1            1300
5343       other      9 BHK     42000.0   8.0  175.0    9             416
5417      Ulsoor      4 BHK     36000.0   4.0  450.0    4            1250
5597    JP Nagar      2 BHK      1100.0   1.0   15.0    2            1363
7166   Yelahanka  1 Bedroom     26136.0   1.0  150.0    1             573
7862    JP Nagar      3 BHK     20000.0   3.0  175.0    3             875
8300     Kengeri      1 BHK      1200.0   1.0   14.0    1            1166
9144       other  4 Bedroom     10961.0   4.0   80.0    4             729
11635      Begur      3 BHK      2400.0   3.0   12.0    3             500
12355      other      4 BHK     16335.0   4.0  149.0    4             912

### When you analyse the above DataFrame, you can see 9 BHK for 416 rupees per SQFt, which is truly an outlier

In [11]:
df[df.price_per_sqft > max_threshold]

location       size  total_sqft  bath   price  bhk  \
345                other  3 Bedroom        11.0   3.0    74.0    3   
1005               other      1 BHK        15.0   1.0    30.0    1   
1106               other  5 Bedroom        24.0   2.0   150.0    5   
4044      Sarjapur  Road  4 Bedroom         1.0   4.0   120.0    4   
4924               other      7 BHK         5.0   7.0   115.0    7   
5911         Mysore Road  1 Bedroom        45.0   1.0    23.0    1   
6356        Bommenahalli  4 Bedroom      2940.0   3.0  2250.0    4   
7012               other      1 BHK       650.0   1.0   500.0    1   
7575               other      1 BHK       425.0   1.0   750.0    1   
7799               other      4 BHK      2000.0   3.0  1063.0    4   
8307   Bannerghatta Road      5 BHK      2500.0   4.0  1400.0    5   
9436        Indira Nagar  4 Bedroom      2400.0   5.0  1250.0    4   
11447         Whitefield  4 Bedroom        60.0   4.0   218.0    4   
12328              other  4 Bedroom      4350.0   8.0  2600.0    4   

       price_per_sqft  
345            672727  
1005           200000  
1106           625000  
4044         12000000  
4924          2300000  
5911            51111  
6356            76530  
7012            76923  
7575           176470  
7799            53150  
8307            56000  
9436            52083  
11447          363333  
12328           59770

In [12]:
cleaned_df = df[(df.price_per_sqft> min_threshold) & (df.price_per_sqft < max_threshold)]
cleaned_df.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_sqft  
0            3699  
1            4615  
2            4305  
3            6245  
4            4250

In [13]:
cleaned_df.sample(10)

location       size  total_sqft  bath   price  bhk  \
626                    other      1 BHK       469.0   1.0   25.00    1   
2558            Haralur Road      2 BHK      1300.0   2.0   73.00    2   
10818                Varthur      2 BHK      1045.0   2.0   40.65    2   
10620        Electronic City      2 BHK      1342.0   2.0   72.00    2   
7293             Brookefield  5 Bedroom      1950.0   6.0  175.00    5   
10561     7th Phase JP Nagar      3 BHK      1976.0   3.0  145.00    3   
5019                   other      2 BHK      1100.0   2.0   47.00    2   
8922                   other      3 BHK      1600.0   3.0  180.00    3   
12515  Raja Rajeshwari Nagar      2 BHK      1157.0   2.0   65.00    2   
7264           Iblur Village      3 BHK      1920.0   3.0  130.00    3   

       price_per_sqft  
626              5330  
2558             5615  
10818            3889  
10620            5365  
7293             8974  
10561            7338  
5019             4272  
8922            11250  
12515            5617  
7264             6770